### DISCUSSION TITLE
Terminal apps open with wrong dimensions when launched with key bindings

### DISCUSSION THREAD

**Discussion by Masacaai** [qtile/qtile#2164] [Q&A]
Hello, I've recently switched from qtile to qtile-git and I noticed that my terminal apps were opening with wrong dimensions when launched. 

I'll use ranger to illustrate my example but this is the same for all my terminal programs:

This is what ranger is supposed to look like when launched. This is what it looked like when launched in qtile (stable) and this is what it looks like in qtile-git when launched manually (without the use of keybindings):

![1611979773](https://user-images.githubusercontent.com/59050442/106346811-344cc500-62d3-11eb-97fc-2f83df63725f.png)

However, when I try to launch ranger through the use of keybindings in qtile-git, this is what I get:

![1611979751](https://user-images.githubusercontent.com/59050442/106346832-6100dc80-62d3-11eb-898b-42a19c3e20ea.png)

This is the relevant part of my config file. Let me know if I need to post more of it:

```
Key([mod], "r",
lazy.spawn("kitty -e ranger"),
desc="Launch ranger"),
```
I haven't made any changes in my config file when switching from qtile to qtile-git. I'm not sure if this is something I can fix myself, but I wanted to confirm before opening a new issue for this.

**Comment by ramnes** [comment]
It has been discussed in the mailing list: https://groups.google.com/g/qtile-dev/c/mjMKB533GNA/m/f0UGhBxpDgAJ

Apparently it is fixed with:
```
COLUMNS=
LINES=
```

That being said, if it used to work without a particular environment, it could mean that we're doing something wrong. Please open an issue and ideally bisect the faulty commit.

**Comment by cjoke** [comment]
I installed alacritty from github, and have the same issue when running alacritty locally, it does not happen while ssh to another computer.

**Reply by cjoke** [reply]
As ramnes mentioned, this fixes issue on my computer running locally.
COLUMNS=
LINES=

This was not an issue for me running urxvt, only on alacritte.

**Comment by 80KiloMett** [comment]
I also have this issue on kitty 0.19.3 and alacritty 0.7.2. On my old st 0.8.4 installation everything works fine however. This also doesn't seem to affect all tui-programs.
In my case ranger and htop are broken. Midnight Commander, vifm, lazygit, vim all work as expected.

Also concerning the workaround:
I had to specifically call `unset` on `COLUMNS` and `LINES` in zsh for it to work.

**Comment by ep12** [comment]
I have that issue, too. For me it's htop and ncmpcpp in scratchpad terminals. Normal windows work fine though. I thought it might be a bug somewhere else (ncurses or some library used by those programs). Glad to see that I'm not alone. But what the heck is going on that this _might_ be qtile's fault? I thought the terminal emulators were calculating the sizes and everything? What has that to do with the WM?

**Reply by ep12** [reply]
I tried different commits: 548a7e040addc9bd8adeac778b2f21c794057e4b is the oldest one I could try; even older commits break my config. The issue persisted with all the commits I checked out. I don't think that this is qtile's fault, but I couldn't find a library used by the terminals I tested that is used by all of the terminal emulators that misbehave. urxvt and roxterm work fine, xterm, qterminal, termite and alacritty screw htop up.

**Reply by Masacaai** [reply]
Oh I just assumed something had changed with the latest qtile release since it worked properly with the previous release.

**Reply by ep12** [reply]
This is totally possible, 548a7e0 is between 0.16.1 and 0.17.0

**Reply by m-col** [reply]
I bisected and found 6a5f3bd781353b0922545f06927a41880adb1bae to be the culprit, though the changes in that commit do not seem related at all. Could someone test that that commit is broken and in the one before it works?

**Reply by ep12** [reply]
Nice! I'll try that later, but I've to set up a config for that as mine is broken before 548a7e0.

**Reply by m-col** [reply]
Same, I used that commit's default config

**Reply by ep12** [reply]
Okay, I modified the default config of either 6a5f3bd781353 or 908b910d00087 (its parent), I can't remember tbh. They shouldn't be different as far as the commit diff 6a5f3bd781353 tells me. The parent seems to be working correctly, indeed:
![before](https://user-images.githubusercontent.com/26854722/111888451-f15ac400-89dc-11eb-9591-a0b3d14f1b12.png)
After checking out 6a5f3bd7813 the environment variables `$LINES` and `$COLUMNS` are somehow magically set, which causes the trouble: `htop` is screwed up:
![after](https://user-images.githubusercontent.com/26854722/111888467-1c451800-89dd-11eb-9a08-943c1bed0a79.png)

The only changes I made to the config were:
+ add a `ScratchPad` containing a `DropDown('htop', 'termite -e test-script', x=0.05, y=0.05, width=0.9, height=0.9)`
+ add a key binding to toggle that drop down
+ add a python script `test-script` to `$PATH`:
   ```py
   #!/usr/bin/env python3
   import os
   import datetime
   import subprocess
   from time import sleep
   if __name__ == '__main__':
   print('os:', os.get_terminal_size())
   print('env:', os.environ.get('COLUMNS', -1), os.environ.get('LINES', -1))
   sleep(1)
   subprocess.check_output(['imlib2_grab', f'/tmp/{datetime.datetime.now().strftime("%Y%m%d%H%M%S")}.png'])
   sleep(10)
   os.execv('/sbin/htop', ['htop'])
   ```

`htop` is not visible in the screen shots, but as far as I know htop is definitely screwed up if the environment variables are set and if they don't match what `os.get_terminal_size()` reports.
Btw: If I run this script in a terminal, it reports the same tuple twice and htop behaves normally.  
Others also experienced the graphical issues this with `termite -e test-script` bound to keys AFAIK. I have not tried that.

Another thing I tried is to reboot with 6a5f3bd7813 checked out. Even directly after a reboot the issue appeared.

---------------------

To conclude that: I think you are right @m-col, 6a5f3bd7813 breaks it. As to why: I have no clue whatsoever.

**Reply by m-col** [reply]
Brilliant, a real head-scratcher this one! I suppose the next step could be to bisect the actual changes within the commit.

**Reply by ep12** [reply]
Hum, yes. But `qshell <~> qtile shell` should never be invoked during nominal operation apart from the user doing it manually I guess? Still not 100 % sure if that really is it. There is no indication that those changes have to do anything with environment variables being set system-wide...
